Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +121,299 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [← toDFA_correct] at hx ⊢
   exact M.toDFA.pumping_lemma hx hlen
 
+/-! ### Regex-like operations
+
+Since NFAs are models for regular languages, they provide decision procedures for the empty set,
+singleton sets with members of at most length 1 as well as a notion of closedness under union,
+concatenation and application of the Kleene star.
+
+Cf. <https://en.wikipedia.org/wiki/Thompson%27s_construction> for the basic idea in a setting
+permitting ε-moves.
+-/
+instance : Zero (NFA α σ) :=
+  ⟨NFA.mk (fun _ _ ↦ ∅) ∅ ∅⟩
+
+instance : Inhabited (NFA α σ) :=
+  ⟨0⟩
+
+@[simp]
+theorem zero_correct : (0 : NFA α σ).accepts = 0 := by
+  ext
+  refine ⟨?_, False.elim⟩
+  rintro ⟨_, ⟨⟩, _⟩
+
+instance : One (NFA α σ) :=
+  ⟨⟨fun _ _ ↦ ∅, univ, univ⟩⟩
+
+@[simp]
+theorem one_correct [Inhabited σ] :
+    (1 : NFA α σ).accepts = 1 := by
+  ext x
+  constructor
+  · rintro ⟨q, _, eval⟩
+    rcases x.eq_nil_or_concat with rfl | ⟨_, _, rfl⟩
+    · exact rfl
+    rw [List.concat_eq_append, eval_append_singleton, mem_stepSet] at eval
+    rcases eval with ⟨_, _, ⟨⟩⟩
+  · rintro rfl
+    exact ⟨default, Set.mem_univ _, Set.mem_univ _⟩
+
+/-- An NFA accepting just a single word of length 1, given by the underlying character. -/
+def char (a : α) : NFA α (σ ⊕ σ') :=
+  ⟨fun p c q ↦ p.isLeft ∧ a = c ∧ q.isRight, (Sum.isLeft ·), (Sum.isRight ·)⟩
+
+@[simp]
+theorem char_correct [Inhabited σ] [Inhabited σ'] (a : α) :
+    (char a : NFA α (σ ⊕ σ')).accepts = {[a]} := by
+  ext x
+  constructor
+  · rintro ⟨_ | _, accept, eval⟩
+    · cases accept
+    rcases x.eq_nil_or_concat with rfl | ⟨as, c, rfl⟩
+    · cases eval
+    rw [List.concat_eq_append, eval_append_singleton, mem_stepSet] at eval
+    rcases eval with ⟨_ | _, mem, step⟩; rotate_left
+    · rcases step with ⟨⟨⟩, _, _⟩
+    rcases as.eq_nil_or_concat with rfl | ⟨_, _, rfl⟩
+    · rcases step with ⟨_, rfl, _⟩; exact rfl
+    rw [List.concat_eq_append, eval_append_singleton, mem_stepSet] at mem
+    rcases mem with ⟨_, _, _, _, ⟨⟩⟩
+  · rintro rfl
+    refine ⟨.inr default, rfl, ?_⟩
+    rw [← List.nil_append [a], eval_append_singleton, mem_stepSet]
+    exact ⟨.inl default, by repeat fconstructor⟩
+
+instance : HAdd (NFA α σ) (NFA α σ') (NFA α (σ ⊕ σ')) :=
+  ⟨fun P₁ P₂ ↦
+    ⟨fun p c q ↦
+      match (p, q) with
+      | (.inl p, .inl q) => P₁.step p c q
+      | (.inr p, .inr q) => P₂.step p c q
+      | _ => False,
+      Sum.elim P₁.start P₂.start,
+      Sum.elim P₁.accept P₂.accept⟩⟩
```

Comment:
I've leaned on the notation for Kleene algebras, where + is the union and * is concatenation. `hadd_correct` and `hmul_correct` illustrate this. Or would you like me to avoid notation in general? I could certainly make `union` and `concat` (or whatever) functions out of these definitions.

Reply:
Heterogeneous addition is so far not used in mathlib, so I am a bit wary of introducing a use of it. I don't want to be dogmatic, so my personal opinion is that removing the notation might be the best thing to do. I would suggest calling the definition `sum` to match the `⊕` that appears in the return type.